platform-sdk/platform-apps/tests/ConsistencyTestingTool/src/main/java/com/swirlds/demo/consistency/TransactionHandlingHistory.java

```diff
@@ -0,0 +1,274 @@
+ * <p>
+ * Contains a record of all rounds that have come to consensus, and the transactions which were included
+ * <p>
+ * Writes a log file containing the history of transaction handling, so that any replayed transactions after a reboot
+ * can be confirmed to match the original handling. NOTE: Partially written log lines are simply ignored by this tool,
+ * so it is NOT verifying handling of transactions in a partially handled round at the time of a crash.
+ */
+public class TransactionHandlingHistory {
+    private static final Logger logger = LogManager.getLogger(TransactionHandlingHistory.class);
+
+    /**
+     * A map from round number to historical rounds
+     */
+    private final Map<Long, ConsistencyTestingToolRound> roundHistory;
+
+    /**
+     * A set of all transactions which have been seen
+     */
+    private final Set<Long> seenTransactions;
+
+    /**
+     * The location of the log file
+     */
+    private Path logFilePath;
+
+    /**
+     * The writer for the log file
+     */
+    private BufferedWriter writer;
+
+    /**
+     * The round number of the previous round handled
+     */
+    private long previousRoundHandled;
+
+    /**
+     * Constructor
+     */
+    public TransactionHandlingHistory() {
+        this.roundHistory = new HashMap<>();
+        this.seenTransactions = new HashSet<>();
+
+        // initialization is happening in init()
+    }
+
+    /**
+     * Initializer
+     * <p>
+     * Reads the contents of the log file if it exists, and adds the included rounds to the history
+     *
+     * @param logFilePath the location of the log file
+     */
+    public void init(final @NonNull Path logFilePath) {
+        this.logFilePath = Objects.requireNonNull(logFilePath);
+
+        logger.info(STARTUP.getMarker(), "Consistency testing tool log path: {}", logFilePath);
+
+        tryReadLog();
+
+        try {
+            this.writer = new BufferedWriter(new FileWriter(logFilePath.toFile(), true));
+        } catch (final IOException e) {
+            throw new RuntimeException("Failed to open writer for transaction handling history", e);
+        }
+    }
+
+    /**
+     * Reads the contents of the log file if it exists, and adds the included rounds to the history
+     */
+    private void tryReadLog() {
+        if (!Files.exists(logFilePath)) {
+            logger.info(STARTUP.getMarker(), "No log file found. Starting without any previous history");
+            return;
+        }
+
+        logger.info(STARTUP.getMarker(), "Log file found. Parsing previous history");
+
+        try (final BufferedReader reader = new BufferedReader(new FileReader(logFilePath.toFile()))) {
+            reader.lines().forEach(line -> {
+                final ConsistencyTestingToolRound parsedRound = ConsistencyTestingToolRound.fromString(line);
+
+                if (parsedRound == null) {
+                    logger.error(EXCEPTION.getMarker(), "Failed to parse line from log file: {}", line);
```

Comment:
is it the `logger.error` that causes validation to fail? Or is it the `EXCEPTION.getMarker()?

Reply:
Currently it is the marker we key off of. A better way to do it would be to key off of `logger.error`... I intend to migrate to this pattern when we eventually gain code ownership of the validators.

Reply:
JVM dying while writing a line will trigger this log. Changing to 

```
logger.warn(STARTUP.getMarker(), "Failed to parse line from log file: {}", line);
```